### PR TITLE
Don't suggest names for errors in normal IDE usage

### LINKED ISSUE #6044: JaroWinkler string distance algorithm allocates 1.11 GB and spends 4.8% of CPU time generating suggestions that are rarely used
This was found by:

* Using VF# solution in VS 2019 Preview
* Working in `service.fs`
* Noticing memory usage was high
* Profiling with good 'ole PerfView
* Typing a bit, invoking some features, etc.

**GC Rollup**
image

Using a 50 second sample of the full trace:

**Allocations**
image

**System.Char[] allocations**
image

**CPU time**
image

There are two problems at play here:

1. We send _all_ identifiers the compiler knows about through this algorithm when generating a set of suggested names when there is a compile error
2. The routine itself allocates a lot of `char[]`s and generally does a lot of work: https://github.com/Microsoft/visualfsharp/blob/master/src/utils/EditDistance.fs#L29

Taking a brief look, it doesn't seem trivial to make this faster or allocate less. It's also not clear how we could generate useful suggestions for people if we cannot scan all known identifiers.

That said, we should find something to do here. It comes as a result of typing in the IDE. Because this is done a lot (duh!), especially for slower typers, improving this would be useful.

### PR DESCRIPTION
**Pull request by cartermp**
Fixes #6044 

This turns off name suggestion for errors in the IDE except for Build, instead using Completion as a source of names for suggesting errors in a Roslyn code fix.

The following will still generate suggestions on build errors:

* fsc.exe
* fsi.exe
* Build within VS
* FSI within VS
* dotnet build

However, names will not be continuously generated when editing code in the IDE and triggering this logic due to having typed an unresolved identifier. This should resolve the current (as of latest dev16.0 branch) biggest source of CPU and memory usage in normal IDE usage for larger solutions.

### CODE REVIEW

**Comment by cartermp** [comment]
After thinking about this a bit more, I think we just have to disable to code fix by default. It's still going to be expensive and depending on how the user mouses over things in the editor will still use lots of CPU and memory.

I don't feel too bad about disabling it entirely, since code fixes are so rarely invoked by users in general.

The real solution here is to have a separate, tree-like structure the IDE builds up and maintains as new identifiers show up, like how Roslyn does it.

File: src/fsharp/service/service.fsi [hunk 1]
```diff
@@ -304,6 +304,10 @@ type public FSharpParsingOptions =
       LightSyntax: bool option
       CompilingFsLib: bool
       IsExe: bool
+
+      /// When true, error messages will include suggestions for parsing errors based on what the compiler knows.
+      /// This is an expensive operation that should be false outside of code fixes, command-line builds, and in-IDE builds.
+      SuggestNamesForErrors: bool
```
Comment (auduchinok) [inline src/fsharp/service/service.fsi]: Do we really need it in parsing options? Suggesting names doesn't happen at parsing stage, does it?

File: src/fsharp/service/service.fsi [hunk 2]
```diff
@@ -352,6 +356,10 @@ type public FSharpProjectOptions =
       /// If two sets of options both have stamps, then they are considered equal
       /// if and only if the stamps are equal
       Stamp: int64 option
+
+      /// When true, error messages will include suggestions for parsing errors based on what the compiler knows.
+      /// This is an expensive operation that should be false outside of code fixes, command-line builds, and in-IDE builds.
+      SuggestNamesForErrors: bool
```
Comment (auduchinok) [inline src/fsharp/service/service.fsi]: I doubt it should be in project options as there's probably no scenario where we want to suggest it for one project or script and not suggest it for others. As this is mostly an environment difference thing (IDE analysis vs build time) I think it should be a `FSharpChecker` parameter instead.

File: src/fsharp/service/service.fsi (hunk 1 above)
Comment (cartermp) [reply src/fsharp/service/service.fsi]: Nope. I'll revisit this on Monday

File: src/fsharp/service/service.fsi (hunk 2 above)
Comment (cartermp) [reply src/fsharp/service/service.fsi]: @auduchinok Yes, that sounds more appropriate. It's probably also a bit less churn. Although having some ability to have it off in one project and on in another (or in script files) may be interesting, having it be a global toggle is consistent with the rest of the IDE features.

**Comment by cartermp** [comment]
I captured my thoughts on that PR but I'll restate the important bit here:

> Fundamentally, this feature is simply _not_ built with IDE tooling in mind. It's great for a batch compile job/command line builds, but for a long-running process like a language server it's simply calculating and re-calculating too many things. The feature needs two "modes":
>
> * The current one for batch compile jobs
> * An out-of-process spellchecker routine that uses a tree (such as [Roslyn's which uses a BKTree](source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Utilities/SpellChecker.cs)) to traverse a populated set of symbols independently of the in-proc work
> 
> The latter is not something that FCS can do yet. We're going to pursue this sort of architecture in the future, but not right now.

Disabling this by default while we are still in the VS process for this sort of work is the right thing to do.

**Comment by forki** [comment]
With the same logic you basically need to disable almost every useful
feature. There is always some cost and recalculation involved.
I mean I understand that this is something that we need to balance - that's
why I really wanted to know the impact of that other PR.

Am Mo., 7. Jan. 2019, 19:12 hat Phillip Carter <notifications@github.com>
geschrieben:

**Comment by cartermp** [comment]
That's a false equivalence.

**Comment by vasily-kirichenko** [comment]
Excellent, but I think fsc.exe should not generate suggestions by default either. It should be a pure-IDE, optional feature.

**Comment by majocha** [comment]
Yes, this looks like the proper way to deal with it.
The code that computes suggestions should be called by the [`FSharpReplaceWithSuggestionCodeFixProvider`](https://github.com/Microsoft/visualfsharp/blob/fead0aac540485683f694524eadad79983ec28d9/vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs#L17) instead.

**Comment by cartermp** [comment]
It'll be heinous to decouple the logic that generates candidates:

https://github.com/Microsoft/visualfsharp/blob/a7acf9f4198550f697945ba36c8321e96541eeab/src/fsharp/NameResolution.fs#L1831-L1838

https://github.com/Microsoft/visualfsharp/blob/a7acf9f4198550f697945ba36c8321e96541eeab/src/fsharp/NameResolution.fs#L1845-L1851

https://github.com/Microsoft/visualfsharp/blob/a7acf9f4198550f697945ba36c8321e96541eeab/src/fsharp/NameResolution.fs#L2209-L2248

https://github.com/Microsoft/visualfsharp/blob/a7acf9f4198550f697945ba36c8321e96541eeab/src/fsharp/NameResolution.fs#L2354-L2392

https://github.com/Microsoft/visualfsharp/blob/a7acf9f4198550f697945ba36c8321e96541eeab/src/fsharp/NameResolution.fs#L2478-L2514

https://github.com/Microsoft/visualfsharp/blob/a7acf9f4198550f697945ba36c8321e96541eeab/src/fsharp/NameResolution.fs#L2577-L2606

https://github.com/Microsoft/visualfsharp/blob/a7acf9f4198550f697945ba36c8321e96541eeab/src/fsharp/NameResolution.fs#L2681-L2696

https://github.com/Microsoft/visualfsharp/blob/a7acf9f4198550f697945ba36c8321e96541eeab/src/fsharp/NameResolution.fs#L2800-L2805

https://github.com/Microsoft/visualfsharp/blob/a7acf9f4198550f697945ba36c8321e96541eeab/src/fsharp/NameResolution.fs#L2821-L2826

https://github.com/Microsoft/visualfsharp/blob/a7acf9f4198550f697945ba36c8321e96541eeab/src/fsharp/NameResolution.fs#L2844-L2852

https://github.com/Microsoft/visualfsharp/blob/a7acf9f4198550f697945ba36c8321e96541eeab/src/fsharp/NameResolution.fs#L2873-L2878

https://github.com/Microsoft/visualfsharp/blob/a7acf9f4198550f697945ba36c8321e96541eeab/src/fsharp/NameResolution.fs#L2885-L2890

https://github.com/Microsoft/visualfsharp/blob/a7acf9f4198550f697945ba36c8321e96541eeab/src/fsharp/NameResolution.fs#L2920-L2933

https://github.com/Microsoft/visualfsharp/blob/a7acf9f4198550f697945ba36c8321e96541eeab/src/fsharp/NameResolution.fs#L3055-L3104

I'm not particularly keen on doing that but I suppose it's possible. Lots and lots of plumbing values to/from NameResolution if we want the code fix to be able to get the same fidelity of results.

A different way could be picked, but something basic like checking matches against all symbols is going to be expensive. If that were crippled to just be within the current document it wouldn't be bad, but then the feature would be inconsistent with using the IDE build.

I'm also _not_ in favor of ripping out this feature for fsc/normal build. It's a valuable feature that I wouldn't be pressing on if it weren't affecting CPU and memory so much in the IDE.

**Comment by dsyme** [comment]
I've noticed this feature to be useful in the IDE.  But I've also noticed the computation to be oddly expensive.

**Comment by forki** [comment]
Even if I disagree with disabling this feature by default: I want to thank you folks for working so hard on perf. I think the community really appreciates that kudos! Also big thanks to @davkean for profiling work.

File: vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs [hunk 3]
```diff
@@ -2,45 +2,62 @@
...
-                        context.RegisterCodeFix(codefix, diagnostics))
-        } |> RoslynHelpers.StartAsyncUnitAsTask(context.CancellationToken)
+        asyncMaybe {
+            do! Option.guard settings.CodeFixes.SuggestNamesForErrors
+
+            let document = context.Document
+            let! _, projectOptions = projectInfoManager.TryGetOptionsForEditingDocumentOrProject(document, context.CancellationToken)
+            let! _, _, checkFileResults = checker.ParseAndCheckDocument(document, projectOptions, userOpName=userOpName)
+            
+            let! allSymbolUses = checkFileResults.GetAllUsesOfAllSymbolsInFile() |> liftAsync
```
Comment (majocha) [inline vsintegration/src/FSharp.Editor/CodeFix/ReplaceWithSuggestion.fs]: Wouldn't it be better to get the list of intellisense completions for the context? Should be fast enough.

**Comment by cartermp** [comment]
@forki Thanks 🙂 - I've actually now changed the code fix to be on by default, but I'd like to play with it a bit more to see what the impact is. This means the code is run differently, so it could mean that the algorithm isn't triggered as much.

**Comment by cartermp** [comment]
@forki Thanks 🙂 - I've actually now changed the code fix to be on by default, but I'd like to play with it a bit more to see what the impact is. This means the code is run differently, so it could mean that the algorithm isn't triggered as much.

**Comment by cartermp** [comment]
As of the current state of this PR (code fix _on_ by default), the allocations in #6044 are practically gone:

image

And CPU usage doesn't even register. So I think this PR solves the problem. However, suggestions are generated differently than with Build. It is equivalent to invoking completion. Is that considered fine?

**Comment by cartermp** [comment]
@forki So unfortunately due to #6036 you won't see ctrl+B unless you toggle "Build Only":

image

We really need to fix that bug for reasons beyond this, but when it is fixed you should probably see both in the error list.

Saving a file does not show it, nor is it in the QuickInfo tooltip (we don't control that, unfortunately). But the fixer icon is in the tooltip area:

image
image

And of course `ctrl+.` will trigger it as well.

**Comment by cartermp** [comment]
@TIHan @KevinRansom this is ready

**Pull request merged**